### DISCUSSION TITLE
Create a collection of "developers" and add GitHub infos to them

### DISCUSSION THREAD

**Discussion by holger1411** [11ty/eleventy#1857] [Q&A] [answered]
Hallo!
Currently I ´am working on a test project and I´am relatively new to Eleventy.
As a test I try to build up a collection of developers and add some infos fetched from the GitHub API to the pages. Now I want to automate this and it seems I need some help with this...

But first things first: [The test project can be found here](https://github.com/holger1411/11ty-api)

I start with a fresh 11ty install. I created a collection subfolder called /developers with an developers.json in it with this content:
```
{
    "layout": "layouts/developer.njk",
    "tags": ["developers"]
}

```
Than I add another file called holger1411.md (<- yep, thats me...) with this content:
```
---
githubHandle: holger1411
---
```

Than I installed node-fetch as dev dependency.

After it I add a /_data folder to the root directory and add the following to it:
```
const fetch = require("node-fetch");
module.exports = async function() {
  console.log("Fetching GitHub data...");

  return fetch("https://api.github.com/users/holger1411")
    .then(res => res.json()) // node-fetch option to transform to json
    .then(json => {
      return {
        url: json.avatar_url,
        twitter: json.twitter_username,
        followers: json.followers
      };
    });
};
```
I use this to fetch some basic infos for my GitHub profile and prepare that data to be available globally.

Okay and than I add a template file called "developer.njk" to the /_includes/layout folder.
That template comes with this content:

```
<!doctype html>
<html>
  <head>
  </head>
  <body>
    <img src="{{ github.url }}" />
    <p>Hi, my Github handle is <a href="https://github.com/{{ githubHandle }}" target="_blank">{{ githubHandle }}</a>
My Twitter handle is {{ github.twitter }} and I have {{ github.followers }}
    </p>
  </body>
</html>
```

And tatatata....it works. It shows my Github profil name, my follower count, my twitter profile handle and my avatar image...

So far so good...but now I want to add another developer to the collection. For example a file called zachleat.md with this content:
```
---
githubHandle: zachleat
---
```

Of course this will render my information instead of zachleat ´s...
At this point I need help. I need a dynamic API call fetching the informations for every GitHub handle in every single markdown files in the "developers" collection and than create pages with the specific informations for each developer...

I´am "just" a UX Designer with limited JS and API knowledge, but I can "smell" already that there is a easy solution near by...
I would be super happy if someone could point me in the right direction.

Big thx upfront!

**Comment by pdehaan** [comment, marked as answer]
I can think of two possible solutions:
- an async shortcode that takes a GitHub username and returns your `img` and `p` tag output.
- modify the global data file to return results for an array of GitHub usernames and use that to create pages using pagination. 

Both solutions have some pros and cons, but guaranteed that you’ll learn something with each approach.

**Reply by pdehaan** [reply]
## Async Nunjucks Shortcode Edition

```js
// .eleventy.js
const fetch = require("node-fetch");

module.exports = eleventyConfig => {
  eleventyConfig.addNunjucksAsyncShortcode("githubProfile", fetchGitHubUser);

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};

async function fetchGitHubUser(githubHandle) {
  console.log(`Fetching GitHub data... ${githubHandle}`);
  const github = await fetch(`https://api.github.com/users/${githubHandle}`)
    .then(res => res.json()) // node-fetch option to transform to json
    .then(json => {
      return {
        url: json.avatar_url,
        twitter: json.twitter_username,
        followers: json.followers
      };
    });

    return `
      <img src="${ github.url }" />
      <p>Hi, my GitHub handle is <a href="https://github.com/${ githubHandle }" target="_blank">${ githubHandle }</a>.</p>
      <p>My Twitter handle is ${ github.twitter } and I have ${ github.followers } followers.</p>
    `;
};
```

Then I have a ./src/developers/holger1411.njk file which calls our GitHub-fetching shortcode:

```nunjucks
---
githubHandle: holger1411
---

{% githubProfile githubHandle %}
```

### OUTPUT

```html
<img src="https://avatars.githubusercontent.com/u/1568158?v=4" />
<p>Hi, my GitHub handle is <a href="https://github.com/holger1411" target="_blank">holger1411</a>.</p>
<p>My Twitter handle is holger1411 and I have 154 followers.</p>
```

Now, if I add a ./src/developers/zachleat.njk with a different `githubHandle` frontmatter, I get this output:

```html
<img src="https://avatars.githubusercontent.com/u/39355?v=4" />
<p>Hi, my GitHub handle is <a href="https://github.com/zachleat" target="_blank">zachleat</a>.</p>
<p>My Twitter handle is null and I have 1390 followers.</p>
```

**NOTE:** looks like not everybody has a Twitter handle defined, so you might need to add more shortcode logic to only show that text in certain cases.

---

### PROS
- Easy to set up.
- Fetching and layout code together (possibly a pro and con). Layout code is also handled in JavaScript strings.

### CONS
- You have to set up one page per profile. Not a big deal if you're only fetching 2-3 pages. Less convenient if you're trying to manage 10+.

**Reply by pdehaan** [reply]
## Pagination Edition

```js
// ./src/_data/developers.js
const fetch = require("node-fetch");

const githubHandles = [
  "holger1411",
  "zachleat"
];

module.exports = async () => {
  const arr = [];
  for (const githubHandle of githubHandles) {
    const profile = await fetchGitHubUser(githubHandle)
    arr.push(profile);
  }
  // Return an array of GitHub profile objects.
  return arr;
}

async function fetchGitHubUser(githubHandle) {
  console.log(`Fetching GitHub data... ${githubHandle}`);
  const github = await fetch(`https://api.github.com/users/${githubHandle}`)
    .then(res => res.json()) // node-fetch option to transform to json
    .then(json => {
      return {
        username: githubHandle,
        url: json.avatar_url,
        twitter: json.twitter_username,
        followers: json.followers
      };
    });
  return github;
};
```


And then we can paginate over that array of data using something like this pagination example:

```njk
---
pagination:
  data: "developers"
  size: 1
  alias: github
permalink: "developers2/{{ github.username | slug }}/"
---

<img src="{{ github.url }}" />
<p>Hi, my GitHub handle is <a href="https://github.com/{{ github.username }}" target="_blank">{{ github.username }}</a>.</p>
<p>My Twitter handle is {{ github.twitter }} and I have {{ github.followers }} followers.</p>
```

### OUTPUT

```html
<!-- developers2/holger1411/index.html -->
<img src="https://avatars.githubusercontent.com/u/1568158?v=4" />
<p>Hi, my GitHub handle is <a href="https://github.com/holger1411" target="_blank">holger1411</a>.</p>
<p>My Twitter handle is holger1411 and I have 154 followers.</p>
```

```html
<!-- developers2/zachleat/index.html -->
<img src="https://avatars.githubusercontent.com/u/39355?v=4" />
<p>Hi, my GitHub handle is <a href="https://github.com/zachleat" target="_blank">zachleat</a>.</p>
<p>My Twitter handle is  and I have 1390 followers.</p>
```

---

### PROS
- Pagination can be more flexible. Add one more GitHub username to a global data file and it will render a new page automatically.
- Data fetching done in JavaScript. Layout code done in whatever language you want (we're using Nunjucks templating here, but you could use Liquid or whatever else). Not limited to JavaScript strings for formatting.

### CONS
- Trickier to set up.

**Reply by Ryuno-Ki** [reply]
It might be even easier … https://twitter.com/zachleat/status/1407749382183735304

**Reply by pdehaan** [reply]
Neat trick!
But I don't think it'd work for getting the twitter handle or follower count (or at least everything I tried gave me HTTP 406 errors).

Although I have been wondering if there would be a way to use the GitHub GraphQL API to get results for multiple users in a single request.

**Reply by holger1411** [reply]
@pdehaan First of all: Thx for the detailed and in-deept help! And the best: **It works!**

I add the shortcode to the template file so I don´t have to repeat it with every content file.
And now I can mimic a Github profile just by adding a .md file with the Profile name inside the Font Matter section.

I add the changes to my demo + a simple frontpage and deploy it to Netlify:
https://angry-golick-09cc78.netlify.app/
I also add a third file called pdehaan.md ;-)

If anyone else want to test it: Fork it, add your own .md file to the collection and open a PR!. Here is the repo again: https://github.com/holger1411/11ty-api

Maybe I will add shortcodes for GitLab, Stackoverflow, dev.to, codepen, Bitbucket etc... so you can have an "aggregated combined social developer profile" ;-)

**Reply by Ryuno-Ki** [reply]
Please sign me up once you have that!
(I was about to query all these APIs myself, since I [signed up to quite some few](https://jaenis.ch/about/) of them)
